src/main/java/it/gov/pagopa/initiative/statistics/service/BaseGenericConsumerService.java

```diff
@@ -0,0 +1,66 @@
+package it.gov.pagopa.initiative.statistics.service;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.springframework.kafka.support.Acknowledgment;
+
+import java.io.IOException;
+import java.util.List;
+
+@Slf4j
+public abstract class BaseGenericConsumerService<E> extends BaseKafkaConsumer<E>{
+    protected BaseGenericConsumerService(String applicationName,
+                                         String consumerGroup,
+                                         ObjectMapper objectMapper) {
+        super(applicationName, consumerGroup, objectMapper);
+    }
+
+    public void evaluate(List<ConsumerRecord<String, String>> records, Acknowledgment acknowledgment, Consumer<?, ?> consumer){
+        log.debug("[{}] Evaluating {} records", getFlowName(), records.size());
+
+        for (ConsumerRecord<String, String> r : records) {
+            try {
+                this.evaluateRecords(r, consumer);
+            } catch (Exception e) {
+                log.error(String.format("Something gone wrong during the evaluation of the payload:%n%s", r.value()), e);
+                this.onRecordError2notify(r,"[%s] Something gone wrong during the evaluation of the payload: %s".formatted(getFlowName(), r.value()), e);
+            }
+        }
+
+        if(acknowledgment!=null){
+            acknowledgment.acknowledge();
+        }
+    }
+
+    /** It will evaluate record*/
+    @SuppressWarnings("java:S3864") // suppressing peek warning: in this case the optimization described will not be performed
+    private void evaluateRecords(ConsumerRecord<String, String> consumerRecord, Consumer<?, ?> consumer){
+        log.debug("[{}] Evaluating record", getFlowName());
+
+        Pair<ConsumerRecord<String, String>, E> record2Payload = null;
+
+        try {
+            record2Payload = Pair.of(consumerRecord, deserialize(consumerRecord.value()));
+        } catch (IOException e) {
+            onDeserializeError(consumerRecord,
+                    "[%s] Unexpected json: %s".formatted(getFlowName(), consumerRecord.value()),
+                    e);
+        }
+
+        if(record2Payload != null && this.isNotRetry(record2Payload)) {
+            evaluate(record2Payload.getRight());
+        }
```

Comment:
if the logic outside the try has to be executed only if the try was successful, let's keep it inside the trye block.

I see you have created a common class where to put common methods (it could be a Utilities class instead, if the core logic is not the same :) )
Moreover, in case of isNotRetry method, you could put into the base class the method accepting just the ConsumerRecord, instead of the Pair, this way here you have not to create an object
```suggestion
        try {
            E payload = deserialize(consumerRecord.value());
            
            if(this.isNotRetry(consumerRecord)) {
                evaluate(payload);
            }
        } catch (IOException e) {
            onDeserializeError(consumerRecord,
                    "[%s] Unexpected json: %s".formatted(getFlowName(), consumerRecord.value()),
                    e);
        }
```